### PR TITLE
clp-s: Implement table packing

### PR DESCRIPTION
**Pull request by gibber9809**
# Description
This PR implements table-packing; we combine small tables together into one compression stream until they reach a certain size threshold in order to avoid having many tiny compression streams. This helps avoid outliers in compression ratio, particularly when we enable features like array-structurization which can create many small table.

On the compression side the key differences are that (1) SchemaWriter now keeps track of the total in-memory size of the table it owns instead of determining it after writing to a compression stream; (2) before compression tables are sorted by that in-memory size, and smaller tables are packed together in sequence until their combined size reaches a certain threshold; and (3) table metadata has been changed to accommodate table packing.

On the decompression side the business of making sure tables are read in the correct order, as well as decompressing the packed streams is implemented in TableReader. Most of the rest of the change is contained in SchemaReader. The logic for reading table metadata is split between TableReader and SchemaReader where TableReader reads metadata about individual compression streams, and SchemaReader reading metadata about how schema tables map to those streams. Also note that schema tables now need to be read in the order they appear in the table metadata, which can be different than schema ID order.

Note: this PR makes the decision to leave uncompressed size of individual schema tables out of the table metadata. This is because uncompressed size can be derived from other metadata we do store, and storing uncompressed size in addition to metadata offsets would actually increase the amount of work we need to do to check an archive isn't corrupt while decompressing it.

# Validation performed
* Validated that this PR fixes bad compression ratio outliers during array structurization
* Validated that performance seems to be within variance compared to before this change

## Summary by CodeRabbit

- **New Features**
	- Added a new `PackedStreamReader` class for improved management of reading packed streams.
	- Introduced a comprehensive metadata schema in the `ArchiveWriter` for better handling of compression streams and schema tables.
	- Added a command-line option for specifying a minimum table size in compression options.

- **Improvements**
	- Enhanced `ArchiveReader` for more efficient schema and metadata management.
	- Updated `ArchiveWriter` to optimize metadata processing and writing.
	- Improved the `SchemaReader` for better data loading and timestamp handling.

- **Bug Fixes**
	- Corrected method calls in `ArchiveReader` and `ArchiveWriter` to align with new schema handling logic.

- **Documentation**
	- Updated method signatures and added new methods to improve clarity and usability.

### CODE REVIEW

File: components/core/src/clp_s/ArchiveWriter.cpp
Comment (wraymo) [inline components/core/src/clp_s/ArchiveWriter.cpp:156]:
```suggestion
    for (auto it = m_id_to_schema_writer.begin(); it != m_id_to_schema_writer.end(); ++it) {
        schemas.push_back(it);
    }
```

File: components/core/src/clp_s/CommandLineArguments.hpp
Comment (wraymo) [inline components/core/src/clp_s/CommandLineArguments.hpp:111]: Should the method name be consistent with the variable name?

File: components/core/src/clp_s/ColumnWriter.hpp
Comment (wraymo) [inline components/core/src/clp_s/ColumnWriter.hpp:124]: Do we need to add this method for other classes?

File: components/core/src/clp_s/ColumnWriter.hpp
Comment (wraymo) [inline components/core/src/clp_s/ColumnWriter.hpp:42]: After the code change, `add_value` will not return any size?

File: components/core/src/clp_s/ColumnWriter.hpp
Comment (wraymo) [inline components/core/src/clp_s/ColumnWriter.hpp:45]: The description is not complete?

File: components/core/src/clp_s/ArchiveWriter.cpp
Comment (wraymo) [inline components/core/src/clp_s/ArchiveWriter.cpp:163]: Do you think we should come up with a better name since in line 203, we use `table_offset`?

File: components/core/src/clp_s/ArchiveWriter.cpp
Comment (wraymo) [inline components/core/src/clp_s/ArchiveWriter.cpp:195]: Maybe we should have a description when we declare these two variables (the format and the usage)?

File: components/core/src/clp_s/ArchiveWriter.cpp
Comment (wraymo) [inline components/core/src/clp_s/ArchiveWriter.cpp:203]: Can we make the order of these four values in the tuple consistent with the writing order?

**Review by wraymo — COMMENTED** [review]
Great work!

File: components/core/src/clp_s/ColumnWriter.hpp
Comment (gibber9809) [reply components/core/src/clp_s/ColumnWriter.hpp:124]: The default implementation that returns 0 works for all of the other column writers. The clp string column writer is the only one that has an extra header to record the size of the encoded variables column.

File: components/core/src/clp_s/ColumnWriter.hpp
Comment (gibber9809) [reply components/core/src/clp_s/ColumnWriter.hpp:42]: It does, but by reference. I'll change it to return by value so that it's less confusing.

File: components/core/src/clp_s/ArchiveWriter.cpp
Comment (gibber9809) [reply components/core/src/clp_s/ArchiveWriter.cpp:162]: I'll change it to current_stream_size, and stream_offset.

File: components/core/src/clp_s/TableReader.hpp
Comment (wraymo) [inline components/core/src/clp_s/TableReader.hpp:86]: Since we use `prev` in other places, do you think it's better change it to `m_prev_table_id`?

**Review by wraymo — COMMENTED** [review]
Great work! One thing to note is that we may need to clarify the distinction between schema and table, as now a table refers to something like a merged table.

**Comment by gibber9809** [comment]
> Great work! One thing to note is that we may need to clarify the distinction between schema and table, as now a table refers to something like a merged table.

Right, yeah that will probably be too confusing to anyone new to this code. We could change the terminology to "streams" or "packed streams" or "packed compression streams" to disambiguate from schema tables, and rename TableReader to PackedStreamReader or something? I kind of like PackedStreamReader since it can hold the double meaning of multiple things packed within a stream, and multiple streams packed together in a file.

We should definitely give it some thought and clear up the terminology before merging at any rate.

**Comment by wraymo** [comment]
> > Great work! One thing to note is that we may need to clarify the distinction between schema and table, as now a table refers to something like a merged table.
> 
> Right, yeah that will probably be too confusing to anyone new to this code. We could change the terminology to "streams" or "packed streams" or "packed compression streams" to disambiguate from schema tables, and rename TableReader to PackedStreamReader or something? I kind of like PackedStreamReader since it can hold the double meaning of multiple things packed within a stream, and multiple streams packed together in a file.
> 
> We should definitely give it some thought and clear up the terminology before merging at any rate.

Yeah, PackedStreamReader is better. Not sure if @kirkrodrigues has any better ideas.

**Comment by kirkrodrigues** [comment]
> Yeah, PackedStreamReader is better. Not sure if @kirkrodrigues has any better ideas.

Seems reasonable.

**Comment by gibber9809** [comment]
This PR should have everything it needs to get merged, so could you take another look @wraymo? If you don't have bandwidth to do the final review soon I'll probably go ahead and do a few more changes to make the metadata format forward-compatible with future encoding plans.

**Comment by gibber9809** [comment]
> This PR should have everything it needs to get merged, so could you take another look @wraymo? If you don't have bandwidth to do the final review soon I'll probably go ahead and do a few more changes to make the metadata format forward-compatible with future encoding plans.

I'm going to go ahead and slightly extend the table packing metadata in the way that I explain in that doc that I shared with you.

File: components/core/src/clp_s/ArchiveWriter.cpp
Comment (wraymo) [inline components/core/src/clp_s/ArchiveWriter.cpp:153]: Can we make it more clear to something like 
```suggestion
* This metadata schema organizes information about packed compression streams and associated 
 * schema tables within a file. The schema is divided into two primary sections:
 * Section 1: Compression Streams Metadata
 * - Contains metadata about each compression stream.
 * - Structure:
 *   - Number of packed streams: <64-bit integer>
 *   - For each stream: 
 *     - Offset into the file: <64-bit integer>
 *     - Uncompressed size: <64-bit integer>
 * - `num_separate_column_schemas` is always 0 in the current implementation.
 * - Undefined section for separate column schemas, reserved for future support.
 * 
 * Section 2: Schema Tables Metadata
 * - Contains metadata about schema tables associated with each compression stream.
 * - Structure:
 *   - Number of schema tables: <64-bit integer>
 *   - For each schema table:
 *     - Stream ID: <64-bit integer>
 *     - Offset into the stream: <64-bit integer>
 *     - Schema ID: <32-bit integer>
 *     - Number of messages: <64-bit integer>
```
and probably move some texts starting from line 155 here

File: components/core/src/clp_s/ArchiveWriter.cpp
Comment (wraymo) [inline components/core/src/clp_s/ArchiveWriter.cpp:168]: Can we reuse `PackedStreamMetadata` here?

**Review by wraymo — APPROVED** [review]

**Pull request merged**
